ldap/servers/slapd/back-ldbm/ldbm_search.c

```diff
@@ -1695,31 +1694,41 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
                     slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
                                   "Bypassing filter test\n");
                     if (ACL_CHECK_FLAG) {
-                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
+                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
                     } else {
                         filter_test = 0;
                     }
-                    if (li->li_filter_bypass_check) {
-                        int ft_rc;
 
+                    /* If we don't check this, we could stomp the filter_test aci denied result. */
+                    if (filter_test == 0 && li->li_filter_bypass_check) {
                         slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry", "Checking bypass\n");
-                        ft_rc = slapi_vattr_filter_test(pb, e->ep_entry, filter,
-                                                        ACL_CHECK_FLAG);
-                        if (filter_test != ft_rc) {
+                        filter_test = slapi_vattr_filter_test(pb, e->ep_entry, filter, 0);
+                        if (filter_test != 0) {
                             /* Oops ! This means that we thought we could bypass the filter test, but noooo... */
                             slapi_log_err(SLAPI_LOG_ERR, "ldbm_back_next_search_entry",
                                           "Filter bypass ERROR on entry %s\n", backentry_get_ndn(e));
-                            filter_test = ft_rc; /* Fix the error */
                         }
                     }
                 } else {
                     /* Old-style case---we need to do a filter test */
-                    filter_test = slapi_vattr_filter_test(pb, e->ep_entry, filter, ACL_CHECK_FLAG);
+                    /*
+                     * IMPORTANT - there is a large and important difference between "filter as intended"
+                     * and filter as executed. The filter as executed is what we optimised to, and this
+                     * can importantly include items like parentid in a one level search. The filter as
+                     * intended however is what the user ASKED for. We shouldn't penalise the user because
+                     * we mucked with their filter, so we check the ACI with the "filter as intended", but
+                     * we need to STILL apply the filter test with "as executed" in case of a test threshold
+                     * shortcut (lest we accidentally prevent the user seeing what they wanted ....)
+                     */
+                    filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
```

Comment:
To be honest I am a bit scared here. Evaluation of the filter is expensive, here instead of one filter matching we may do two matching. So the gain on candidate list may be lost into a loss in more expensive filter matching.

Reply:
The whole theory of the filter test threshold is that filter eval is faster than evaling idls because we don't have an idl cache so they are always IO bound. That's why we have the ability to shortcut out and then evaluate the remaining conditions of the filter on the entry rather than doing full-indexed queries and bypassing the filter test. 


I'm actually reading the current master branch code and I have no idea how this it actually working correctly for current queries. 

Looking at this we have `onelevel_candidates()` which wraps the filter as intended by the user, with `(&(filter_intended)(parendid=<id>))`.  Lets say the filter_intended is `uid=a`. We then call into filter_candidates, which will look into the first elements of the filter_indended, and see it's equality, we'll hit the test_threshold and then return out with an idlist of just uid=a,. 

But then in onelevel, we discard the wrapper filter, and return the flag for forcing the filter test. So we then filter test only on filter_intended. 

I'm trying to think of ways to corrupt this, because I think it may be possible to ask for a query that would have subtree elements accidentally included in our partial candidate set since we never actually evaluated parentid=, and we don't apply that condition in our filter test.


So that's why we MUST absolutely test the filter twice! We have to check "once" for the access controls of "what did the user intend" and a second of "What is the correct *query* that gets the right results". Both onelevel and referrals matter here because the user may not have read on objectClass or parentid, and because we inject both, we need to check these are properly in the result sets else we are excluding/over-including in our results.

Reply:
> The whole theory of the filter test threshold is that filter eval is faster than evaling idls because we don't have an idl cache so they are always IO bound. That's why we have the ability to shortcut out and then evaluate the remaining conditions of the filter on the entry rather than doing full-indexed queries and bypassing the filter test.

filter threshold is to interrupt the process of building (from the filter) the candidate list when the #candidate is low (10 IIRC) and move immediately to filter evaluation. IIRC the main issue with filter threshold is when we have a 'OR' that force to fetch candidates for all filter components.